File: core/src/galenscovell/oregontrail/map/Sector.java
package galenscovell.oregontrail.map;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.*;
import com.badlogic.gdx.scenes.scene2d.*;
import com.badlogic.gdx.scenes.scene2d.utils.ActorGestureListener;
import galenscovell.oregontrail.util.*;

public class Sector extends Actor {
    public final int x, y;
    private int frames;
    private SectorType type;
    private Sprite sprite;
    private boolean glowUp;

    public Sector(int x, int y) {
        this.x = x;
        this.y = y;
        this.frames = 60;
        this.glowUp = true;
        becomeEmpty();

        this.addListener(new ActorGestureListener() {
            public void touchDown(InputEvent event, float x, float y, int pointer, int button) {
                if (!isEmpty() && !isExplored()) {
                    Repository.setSelection(getThisSector());
                }
            }
        });
    }

    public Sector getThisSector() {
        return this;
    }

    public boolean isEmpty() {
        return type == SectorType.EMPTY;
    }

    public boolean isCurrent() {
        return type == SectorType.CURRENT;
    }

    public boolean isExplored() {
        return type == SectorType.EXPLORED;
    }

    public boolean isUnexplored() {
        return type == SectorType.UNEXPLORED;
    }

    public void becomeEmpty() {
        type = SectorType.EMPTY;
    }

    public void becomeCurrent() {
        ResourceManager.currentMarker.setTarget(
                x  * Constants.SECTORSIZE - Constants.SECTORSIZE,
                Gdx.graphics.getHeight() - (y * Constants.SECTORSIZE) - (4 * Constants.SECTORSIZE)
        );
        type = SectorType.CURRENT;
    }

    public void becomeExplored() {
        type = SectorType.EXPLORED;
    }

    public void becomeUnexplored() {
        sprite = ResourceManager.sp_test0;
        type = SectorType.UNEXPLORED;
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        if (!isEmpty()) {
            drawGlow(batch);
            batch.draw(
                    sprite,
                    x * Constants.SECTORSIZE,
                    Gdx.graphics.getHeight() - (y * Constants.SECTORSIZE) - (3 * Constants.SECTORSIZE),
                    Constants.SECTORSIZE,
                    Constants.SECTORSIZE
            );
            if (isCurrent()) {
                ResourceManager.currentMarker.render(batch);
            }
        }
    }

    private void drawGlow(Batch batch) {
        if (glowUp) {
            frames++;
        } else {
            frames -= 2;
        }

        if (frames == 120) {
            glowUp = false;
        } else if (frames == 40) {
            glowUp = true;
        }

        float frameAlpha = (frames / 120.0f);

        if (isExplored()) {
            batch.setColor(0.4f, 0.4f, 1.0f, frameAlpha);
        } else {
            batch.setColor(0.4f, 1.0f, 0.4f, frameAlpha);
        }

        batch.draw(
                ResourceManager.mapGlow,
                x * Constants.SECTORSIZE - Constants.SECTORSIZE,
                Gdx.graphics.getHeight() - (y * Constants.SECTORSIZE) - (4 * Constants.SECTORSIZE),
                Constants.SECTORSIZE * 3,
                Constants.SECTORSIZE * 3
        );
        batch.setColor(1, 1, 1, 1);
    }
}

File: core/src/galenscovell/oregontrail/map/Sector.scala
package galenscovell.oregontrail.map

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d._
import com.badlogic.gdx.scenes.scene2d._
import com.badlogic.gdx.scenes.scene2d.utils.ActorGestureListener
import galenscovell.oregontrail.util._


class Sector(x: Int, y: Int) extends Actor {
  val sx: Int = x
  val sy: Int = y
  private var frames: Int = 60
  private var sectorType: Short = 0
  private var sprite: Sprite = null
  private var glowing: Boolean = true

  this.addListener(new ActorGestureListener() {
    override def touchDown(event: InputEvent, x: Float, y: Float, pointer: Int, button: Int): Unit = {
      if (!isEmpty) {
        Repository.setSelection(getThisSector)
      }
    }
  })


  def getThisSector(): Sector = {
    this
  }

  def isEmpty(): Boolean = {
    sectorType == Constants.SECTOR_EMPTY
  }

  def isCurrent(): Boolean = {
    sectorType == Constants.SECTOR_CURRENT
  }

  def isExplored(): Boolean = {
    sectorType == Constants.SECTOR_EXPLORED
  }

  def isUnexplored(): Boolean = {
    sectorType == Constants.SECTOR_UNEXPLORED
  }

  def becomeEmpty(): Unit = {
    sectorType = Constants.SECTOR_EMPTY
  }

  def becomeCurrent(): Unit = {
    ResourceManager.currentMarker.setTarget(sx * Constants.SECTORSIZE - Constants.SECTORSIZE, Gdx.graphics.getHeight - (sy * Constants.SECTORSIZE) - (4 * Constants.SECTORSIZE))
    sectorType = Constants.SECTOR_CURRENT
  }

  def becomeExplored(): Unit = {
    sectorType = Constants.SECTOR_EXPLORED
  }

  def becomeUnexplored(): Unit = {
    sprite = ResourceManager.sp_test0
    sectorType = Constants.SECTOR_UNEXPLORED
  }

  override def draw(batch: Batch, parentAlpha: Float): Unit = {
    if (!isEmpty) {
      glow(batch)
      batch.draw(sprite, sx * Constants.SECTORSIZE, Gdx.graphics.getHeight - (sy * Constants.SECTORSIZE) - (3 * Constants.SECTORSIZE), Constants.SECTORSIZE, Constants.SECTORSIZE)
      if (isCurrent) {
        ResourceManager.currentMarker.render(batch)
      }
    }
  }

  private def glow(batch: Batch): Unit = {
    if (glowing) {
      frames += 1
    }
    else {
      frames -= 2
    }

    if (frames == 120) {
      glowing = false
    }
    else if (frames == 40) {
      glowing = true
    }

    val frameAlpha: Float = (frames / 120.0f)
    if (isExplored) {
      batch.setColor(0.4f, 0.4f, 1.0f, frameAlpha)
    }
    else {
      batch.setColor(0.4f, 1.0f, 0.4f, frameAlpha)
    }

    batch.draw(ResourceManager.mapGlow, sx * Constants.SECTORSIZE - Constants.SECTORSIZE, Gdx.graphics.getHeight - (sy * Constants.SECTORSIZE) - (4 * Constants.SECTORSIZE), Constants.SECTORSIZE * 3, Constants.SECTORSIZE * 3)
    batch.setColor(1, 1, 1, 1)
  }
}
